### ISSUE TITLE
Typings for __typename is wrong in interface resolver

### ISSUE THREAD

**Issue by daeschlerdavid** [graphql-nexus/nexus#866]
Hello

I am implementing a Node interface with `DMF strategy`.
Typings is telling me that the root object have a `__typename` field :

![image](https://user-images.githubusercontent.com/2098583/111879923-dbcba700-89a8-11eb-85f8-f4cfbfce9a6a.png)

I am explicitly adding the `__typename` field when required (when my resolver is returning a Node type)
But in others explicit type resolvers, I tought the `__typename` was automaticaly added. But it is not and __typename is `undefined`.

So I have to return `__typename` in every resolvers ?
Or are the typings wrong ?

**Comment by jasonkuhrt** [comment]
> I tought the __typename was automaticaly added.

Are you saying you expected the Nexus to somehow add `__typename` for you at runtime? This is not the case.

**Comment by daeschlerdavid** [comment]
Yes but then why are the typings saying that the __typename field is available ?
It's optionnal in resolvers, so the typings should be "Tag" | ... | undefined

**Comment by jasonkuhrt** [comment]
The interface resolver source types are a union of all interface implementors. 

For resolvers where __type name is optional, can you show an example?

**Comment by jasonkuhrt** [comment]
Also aside: root is a confusing name. Use "source" instead, or "model" if that fits your application better.

**Comment by tgriesser** [comment]
> Are you saying you expected the Nexus to somehow add __typename for you at runtime? This is not the case.

I wonder if this could be implemented as a plugin 🤔

**Comment by daeschlerdavid** [comment]
Consider this code :

```ts

export const Node2Interface = interfaceType({
  name: 'Node2',
  definition(t) {
    t.nonNull.id('id', {
      resolve(source, args, ctx) {
        // here, "source.__typename" is type safe regarding the typings but it can be undefined
        console.log(source.__typename)

        return 'hello'
      }
    })
  }
})

export const User2Type = objectType({
  name: 'User2',
  definition(t) {
    t.implements('Node2')
    t.nonNull.string('username')
  }
})

export const getUser2 = extendType({
  type: 'Query',
  definition(t) {
    t.field('getUser2', {
      type: 'User2',
      async resolve(source, args, ctx) {
        return {
          // __typename: 'User2', // <-- here __typename is optionnal
          username: 'user'
        }
      }
    })
  }
})

export const getNode2 = extendType({
  type: 'Query',
  definition(t) {
    t.field('getNode2', {
      type: 'Node2',
      async resolve(source, args, ctx) {
        return {
          __typename: 'User2', // <-- here __typename is mandatory
          username: 'user'
        }
      }
    })
  }
})

```

Returning the __typename for a Node2 interface is mandatory.
Returning the __typename for a User2 type is not required because the type is explicit.

I think its related to this in the documentation :

> Nexus is smart about what it needs to be satisfied by __typename presence. Rather than being a requirement of the model type, it is a requirement of the model type in resolver cases under the union type. 

The problem is when you want to use to __typename field in the resolver of the Node2 interface:

Querying this works :

```
{
  getNode2 {
    id
    __typename
    ... on User2 {
      username
    }
  }
}
```

console.log(source.__typename) is printing "User2"

But with this query :

```
{
  getUser2 {
    id
    username
    __typename
  }
}
```

console.log(source.__typename) is printing "undefined"

This is strange because the __typename of the http response is correctly returning "User2".
So it means that nexus is aware of the typename but it is not available to use in the id field resolver.

I think its a common pattern with Relay to hide the id implementation by prefixing it with the typename like : 
id = base64(type:id)

This is what I was trying to achieve but I was surprised to see an undefined field beside the fact that the typings is saying that __typename is safe to use.

I solved this by adding this to my sources types : 

```
export type Actor = prisma.Actor & { __typename: 'Actor' }
export type User = prisma.User & { __typename: 'User' }
```

This way I am forced to return the __typename no matter the type I am resolving.

Now __typename is no more undefined.

**Comment by tgriesser** [comment]
> This way I am forced to return the __typename no matter the type I am resolving.

Yeah, this is the assumption if you're using DMF. I'd have to check on why we're not checking that. I wonder if we should to add a wrapper on the generated types which intersects the typename as you're doing manually there.

e.g. 

```ts
type DMF<TypeName, TypeShape> = TypeShape & { __typename: 'Project' }

Project: DMF<'Project', {
  // root type
  project?: string | null; // String
}>;
```

@jasonkuhrt wdyt?

**Comment by xiavn** [comment]
I've recently ran into this issue myself - I'm currently playing around trying to delve into the world of graphql, and am trying to rustle up a test project before attempting to use anything in production. I've been using a node interface to try and create a relay style schema and this post put me onto the DMF strategy.

@tgriesser  I think a wrapper around the types like you are suggesting would definitely be helpful for helping to resolve this issue, as I've been kind of muddling through trying to understand where I need to add the __typename property and where I do not - well just muddling through in general to be honest. I guess the biggest downside is that by implementing it this way, you might be forced to add the property in situations where it isn't needed?

**Comment by jasonkuhrt** [comment]
@tgriesser off hand that sounds promising but I'm pretty far removed from this feature at this point, would need to dive back in
